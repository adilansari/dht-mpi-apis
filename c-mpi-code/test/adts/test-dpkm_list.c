
/**
 * Test dpkm_list functionality.
 * */

#include <dpkm_list.h>

#include "test_helpers.h"

void extra(struct dpkm_list_item* item)
{
  if (item)
  {
    printf("extra: %s\n", item->key);
    free(item);
  }
}

void extra_data(struct dpkm_list_item* item)
{
  if (item)
  {
    printf("extra: %s\n", item->key);
    printdata("data: ", item->data, strlen(item->data));
    printf("\n");
    free(item->data);
    free(item);
  }
}

void extra_key_data(struct dpkm_list_item* item)
{
  if (item)
  {
    printf("extra: %s\n", item->key);
    printdata("data: ", item->data, item->length);
    printf("\n");
    free(item->key);
    free(item->data);
    free(item);
  }
}

int
main(int argc, char* argv[])
{
  struct dpkm_list* L = dpkm_list_create(4);

  char s[1000];

  /// Begin tests with stack data...
  int zero  = 0;
  int two   = 2;
  int three = 3;
  int four  = 4;
  int five  = 5;
  int eight = 8;

  struct dpkm_list_item* item;

  item = dpkm_list_add(L, "two", &two, sizeof(int));
  extra(item);
  item = dpkm_list_add(L, "four", &four, sizeof(int));
  extra(item);
  item = dpkm_list_add(L, "three", &three, sizeof(int));
  extra(item);
  item = dpkm_list_add(L, "three", &three, sizeof(int));
  extra(item);
  item = dpkm_list_add(L, "four", &four, sizeof(int));
  extra(item);
  item = dpkm_list_add(L, "zero", &zero, sizeof(int));
  extra(item);

  dpkm_list_printf("%i", L);
  printf("size: %i \n", L->size); // 4

  item = dpkm_list_add(L, "five", &five, sizeof(int));
  extra(item);
  item = L->head;
  dpkm_list_excise(L, L->head);
  extra(item);
  dpkm_list_printf("%i", L);
  printf("size: %i \n", L->size); // 3

  item = dpkm_list_add(L, "eight", &eight, sizeof(int));
  extra(item);
  dpkm_list_printf("%i", L);
  printf("size: %i \n", L->size); // 4

  dpkm_list_tostring(s, 1000, "%i", L);
  printf("From string: %s\n", s);

  item = dpkm_list_poll(L);
  extra(item);
  dpkm_list_printf("%i", L);
  printf("size: %i \n", L->size); // 3

  item = dpkm_list_search(L, "zero");
  dpkm_list_excise(L, item);
  extra(item);
  dpkm_list_printf("%i", L);

  puts("polling...");

  while ((item = dpkm_list_poll(L)))
  {
    printf("got: %s \n", item->key);
    extra(item);
    dpkm_list_printf("%i", L);
  }

  dpkm_list_tostring(s, 1000, "%i", L);
  printf("From string: %s\n", s);

  assert(L->size == 0);
  dpkm_list_free(L);

  struct dpkm_list* L2 = dpkm_list_create(4);
  char* binary = heap("data");
  dpkm_list_add(L2, "binary", binary, strlen(binary)+1);
  dpkm_list_printdata(L2);

  item = dpkm_list_poll(L2);
  extra_data(item);

  for (int i = 0; i < 30; i++)
  {
    int j = i % 10;
    if (rand() % 2)
    {
      dpkm_list_printdata(L2);
      char* key = makekey(j);
      void* data = makedata(j);
      printf("adding: %s\n", key);
      item = dpkm_list_add(L2, key, data, j);
      extra_key_data(item);
    }
  }
  dpkm_list_printdata(L2);
  printf("size: %i\n", L2->size);

  while ((item = dpkm_list_poll(L2)))
    extra_key_data(item);

  printf("size: %i\n", L2->size);
  assert(L2->size == 0);
  dpkm_list_free(L2);
}
