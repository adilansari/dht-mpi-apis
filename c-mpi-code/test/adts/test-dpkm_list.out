extra: two
extra: four
[(three,3),(three,3),(four,4),(zero,0)]
size: 4 
extra: three
extra: three
[(four,4),(zero,0),(five,5)]
size: 3 
[(four,4),(zero,0),(five,5),(eight,8)]
size: 4 
From string: [(four,4),(zero,0),(five,5),(eight,8)]
extra: four
[(zero,0),(five,5),(eight,8)]
size: 3 
extra: zero
[(five,5),(eight,8)]
polling...
got: five 
extra: five
[(eight,8)]
got: eight 
extra: eight
[]
From string: []
[(binary@5,[data\0])]
extra: binary
data: [data]
[]
adding: 0
[(0@0,[])]
adding: 2
[(0@0,[]),(2@2,[XX])]
adding: 3
[(0@0,[]),(2@2,[XX]),(3@3,[XXX])]
adding: 4
[(0@0,[]),(2@2,[XX]),(3@3,[XXX]),(4@4,[XXXX])]
adding: 5
extra: 0
data: []
[(2@2,[XX]),(3@3,[XXX]),(4@4,[XXXX]),(5@5,[XXXXX])]
adding: 8
extra: 2
data: [XX]
[(3@3,[XXX]),(4@4,[XXXX]),(5@5,[XXXXX]),(8@8,[XXXXXXXX])]
adding: 9
extra: 3
data: [XXX]
[(4@4,[XXXX]),(5@5,[XXXXX]),(8@8,[XXXXXXXX]),(9@9,[XXXXXXXXX])]
adding: 1
extra: 4
data: [XXXX]
[(5@5,[XXXXX]),(8@8,[XXXXXXXX]),(9@9,[XXXXXXXXX]),(1@1,[X])]
adding: 3
extra: 5
data: [XXXXX]
[(8@8,[XXXXXXXX]),(9@9,[XXXXXXXXX]),(1@1,[X]),(3@3,[XXX])]
adding: 4
extra: 8
data: [XXXXXXXX]
[(9@9,[XXXXXXXXX]),(1@1,[X]),(3@3,[XXX]),(4@4,[XXXX])]
adding: 0
extra: 9
data: [XXXXXXXXX]
[(1@1,[X]),(3@3,[XXX]),(4@4,[XXXX]),(0@0,[])]
adding: 2
extra: 1
data: [X]
[(3@3,[XXX]),(4@4,[XXXX]),(0@0,[]),(2@2,[XX])]
adding: 3
extra: 3
data: [XXX]
[(4@4,[XXXX]),(0@0,[]),(2@2,[XX]),(3@3,[XXX])]
adding: 7
extra: 4
data: [XXXX]
[(0@0,[]),(2@2,[XX]),(3@3,[XXX]),(7@7,[XXXXXXX])]
adding: 8
extra: 0
data: []
[(2@2,[XX]),(3@3,[XXX]),(7@7,[XXXXXXX]),(8@8,[XXXXXXXX])]
adding: 9
extra: 2
data: [XX]
[(3@3,[XXX]),(7@7,[XXXXXXX]),(8@8,[XXXXXXXX]),(9@9,[XXXXXXXXX])]
size: 4
extra: 3
data: [XXX]
extra: 7
data: [XXXXXXX]
extra: 8
data: [XXXXXXXX]
extra: 9
data: [XXXXXXXXX]
size: 0
